test/e2e/storage/volume_io.go

```diff
@@ -0,0 +1,890 @@
+	if err != nil {
+		return fmt.Errorf("client pod %q not running: %v", clientPod.Name, err)
+	}
+
+	// create files of the passed-in file sizes and verify test file size and content
+	for _, fsize := range fsizes {
+		// file sizes must be a multiple of `minFileSize`
+		if math.Mod(float64(fsize), float64(minFileSize)) != 0 {
+			fsize = fsize/minFileSize + minFileSize
+		}
+		fpath := path.Join(dir, fmt.Sprintf("%s-%d", file, fsize))
+		if err = writeToFile(clientPod, fpath, dd_input, fsize); err != nil {
+			return err
+		}
+		if err = verifyFile(clientPod, fpath, fsize, dd_input); err != nil {
+			return err
+		}
+		deleteFile(clientPod, fpath)
+	}
+
+	return
+}
+
+// These tests need privileged containers which are disabled by default.
+var _ = framework.KubeDescribe("Volume plugin streaming [Volume][Slow]", func() {
+	f := framework.NewDefaultFramework("volume-io")
+	var (
+		config    framework.VolumeTestConfig
+		cs        clientset.Interface
+		ns        string
+		serverIP  string
+		serverPod *v1.Pod
+		volSource v1.VolumeSource
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace.Name
+	})
+
+	////////////////////////////////////////////////////////////////////////
+	// NFS
+	////////////////////////////////////////////////////////////////////////
+	framework.KubeDescribe("NFS", func() {
+		testFile := "nfs_io_test"
+		// client pod uses selinux
+		podSec := v1.PodSecurityContext{
+			SELinuxOptions: &v1.SELinuxOptions{
+				Level: "s0:c0,c1",
+			},
+		}
+
+		BeforeEach(func() {
+			config = framework.VolumeTestConfig{
+				Namespace:   ns,
+				Prefix:      "nfs",
+				ServerImage: framework.NfsServerImage,
+				ServerPorts: []int{2049},
+			}
+			serverPod = framework.StartVolumeServer(cs, config)
+			serverIP = serverPod.Status.PodIP
+			framework.Logf("NFS server IP address: %v", serverIP)
+
+			volSource = v1.VolumeSource{
+				NFS: &v1.NFSVolumeSource{
+					Server:   serverIP,
+					Path:     "/",
+					ReadOnly: false,
+				},
+			}
+		})
+
+		AfterEach(func() {
+			framework.Logf("AfterEach: deleting NFS server pod %q...", serverPod.Name)
+			err := framework.DeletePodWithWait(f, cs, serverPod)
+			Expect(err).NotTo(HaveOccurred(), "AfterEach: NFS server pod failed to delete")
+		})
+
+		It("should write files of various sizes, verify size, validate content", func() {
+			fileSizes := []int64{1 * framework.MiB, 100 * framework.MiB, 1 * framework.GiB}
+			err := testVolumeIO(f, cs, config, volSource, &podSec, testFile, fileSizes)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	////////////////////////////////////////////////////////////////////////
+	// Gluster
+	////////////////////////////////////////////////////////////////////////
+	framework.KubeDescribe("GlusterFS", func() {
+		var (
+			endpoints *v1.Endpoints
+			name      string
+		)
+		testFile := "gluster_io_test"
+
+		BeforeEach(func() {
+			framework.SkipUnlessNodeOSDistroIs("gci")
+			var err error
+			config = framework.VolumeTestConfig{
```

Comment:
Should we add test wrappers for starting all these different storage servers?  I believe we already have some for nfs?